requirements.txt

```diff
@@ -1,156 +1,201 @@
-appdirs==1.4.0
-astroid==1.4.8
-Babel==2.3.4
-backports.shutil-get-terminal-size==1.0.0
-beautifulsoup4==4.5.1
-boto3==1.9.64
-botocore==1.12.64
-chardet==2.3.0
-CherryPy==3.8.2
-click==6.6
-cliff==2.3.0
-cmd2==0.6.9
-colorama==0.3.7
+ansible==2.3.2.0
+appdirs==1.4.3
+asn1crypto==0.24.0
+astroid==2.1.0
+attrs==19.1.0
+Babel==2.7.0
+backcall==0.1.0
+backports.functools-lru-cache==1.5
+bcrypt==3.1.4
+beautifulsoup4==4.7.1
+boto==2.48.0
+boto3==1.9.174
+botocore==1.12.174
+certifi==2019.6.16
+cffi==1.12.3
+chardet==3.0.4
+cheroot==6.5.5
+CherryPy==18.1.1
+Click==7.0
+cliff==2.15.0
+cmd2==0.9.13
+colorama==0.4.1
 colour==0.1.5
 cookies==2.2.1
-coverage==4.2
-ddt==1.1.1
-debtcollector==1.9.0
-decorator==4.0.10
-Django==1.11.20
--e git+https://github.com/jrief/django-angular.git@3a4eb9e716b0f03e4abe5ec60d6978a52da100f8#egg=django_angular
-django-appconf==1.0.2
-django-compressor==2.1
-django-debug-toolbar==1.6
-django-environ==0.4.0
-django-extensions==1.7.4
-django-filter==0.15.3
-django-grappelli==2.8.3
+coverage==4.5.3
+cryptography==2.7
+ddt==1.2.1
+debtcollector==1.21.0
+decorator==4.4.0
+Django==2.2.2
+# with Django 1 we were using: -e git+https://github.com/jrief/django-angular.git@3a4eb9e716b0f03e4abe5ec60d6978a52da100f8#egg=django_angular	
+# with Django 2, clean option:
+# django-angular==2.2
+# but django-angular 2.2 requires a heavy rewrite of many parts of the form. Disabling django-angular does too. So a compromise is to patch our django-angular version to work with Django 2:
+# TODO: use opencraft's branch
+-e git+https://github.com/open-craft/django-angular.git@clemente/migrate-our-django-angular-to-django-2#egg=django_angular
+django-appconf==1.0.3
+django-compressor==2.3
+django-debug-toolbar==2.0
+django-environ==0.4.5
+django-extensions==2.1.9
+django-filter==2.1.0
+django-fine-uploader==0.2.0
+django-grappelli==2.12.3
 django-libsass==0.7
-django-redis==4.6.0
-django-simple-email-confirmation==0.21
-django-storage-swift==1.2.17
+django-redis==4.10.0
+django-simple-email-confirmation==0.60
+django-storage-swift==1.2.19
 django-zurb-foundation==5.5.0
-djangorestframework==3.5.3
+djangorestframework==3.9.4
 docutils==0.14
 dodgy==0.1.9
-factory-boy==2.7.0
-fake-factory==0.7.2
-freezegun==0.3.8
+dogpile.cache==0.7.1
```

Comment:
There appear to be some dependencies here that aren't in the original requirements.txt. For example, dogpile is new, and I can't find it directly used anywhere in the codebase. If they are dependencies of dependencies, then perhaps they shouldn't be listed here.

Reply:
They are indeed dependencies of dependencies. I created [this dependency tree](https://gist.github.com/clemente/4d0a3dfd0337e4259a12fda6633978ae) with pipdeptree, and there we see that some are several levels deep.

I think they should be listed, mainly to allow us to do a simple `pip freeze > requirements.txt` and not have to manually edit a lot of information afterwards. And 2nd, to pin them to a version that works; this is to avoid auto-upgrading and to solve possible conflicts. Dependencies of dependencies can still create conflicts (A requires P v1, B requires P v2).

I initially deleted everything that we weren't directly using, and in the end I gave up since it was much more complex to maintain than using `pip freeze`.

Reply:
I'm not a huge fan because it makes it difficult to see at a glance which are our pinned dependencies, and which are random sub dependencies. Also blindly using the output of pip freeze can result in unwanted packages making it into the deps list because of previously installing temporary packages for whatever reason. For example, you have added `pipdeptree`, and `pip-autoremove` which would have been in your local environment but I'm pretty sure aren't used anywhere in ocim.

Reply:
I know and that's how this file has been maintained until now. We had a task to simplify this. I answered in SE-713.
I found it wouldn't be too hard to migrate to a cleaner system so I just did it, to pip-tools (i.e. pip-compile), and now we have requirements.in with the basic dependencies and it autogenerates requirements.txt with all of them. I also removed the packages that we seem we're not using (there were lots more).

Reply:
Nice! At the very least this provides a way to tell the difference between packages we care about and sub dependencies. :+1: